dynacred/src/Contact.ts

```diff
@@ -123,82 +135,98 @@ export class Contact {
             this.incVersion();
         }
     }
 
+    get joinedChallenge(): Long {
+        const longArray = this.credential.getAttribute("1-public", "challenge");
+        return Long.fromBytesLE(Array.from(longArray));
+    }
+
+    set joinedChallenge(v: Long) {
+        if (!this.joinedChallenge.equals(v)) {
+            this.credential.setAttribute("1-public", "challenge", Buffer.from(v.toBytesLE()));
+            this.incVersion();
+        }
+    }
+
     get personhoodPub(): Public | undefined {
         const raw = this.credential.getAttribute("1-public", "personhood");
         if (raw === undefined) {
             return undefined;
         }
-        if (raw === null) {
-            Log.warn(`Contact("${this.alias}").credential has null 1-public/personhood`);
-            return undefined;
-        }
         return Public.fromBuffer(raw);
     }
 
     set personhoodPub(pub: Public) {
-        if (pub) {
+        if (this.personhoodPub === undefined ||
+                !pub.equal(this.personhoodPub)) {
             this.credential.setAttribute("1-public", "personhood", pub.toBuffer());
             this.incVersion();
         }
     }
 
-    get coinID(): InstanceID {
+    get coinID(): InstanceID | undefined {
         return this.credential.getAttribute("1-public", "coin");
     }
 
     set coinID(id: InstanceID) {
-        if (id) {
+        if (this.coinID === undefined || Buffer.compare(this.coinID, id) !== 0) {
             this.credential.setAttribute("1-public", "coin", id);
             this.incVersion();
         }
     }
 
-    get ltsID(): InstanceID {
+    get ltsID(): InstanceID | undefined {
         return this.credential.getAttribute("1-config", "ltsID");
     }
 
     set ltsID(id: InstanceID) {
-        if (id) {
+        if (this.ltsID === undefined || !id.equals(this.ltsID)) {
             this.credential.setAttribute("1-config", "ltsID", id);
             this.incVersion();
         }
     }
 
-    get ltsX(): Point {
+    get ltsX(): Point | undefined {
         const lx = this.credential.getAttribute("1-config", "ltsX");
-        return lx ? PointFactory.fromProto(lx) : null;
+        return lx ? PointFactory.fromProto(lx) : undefined;
     }
 
     set ltsX(X: Point) {
-        if (X) {
+        if (this.ltsX === undefined || !X.equals(this.ltsX)) {
             this.credential.setAttribute("1-config", "ltsX", X.toProto());
             this.incVersion();
         }
     }
 
-    get spawnerID(): InstanceID {
+    get spawnerID(): InstanceID | undefined {
         return this.credential.getAttribute("1-config", "spawner");
     }
 
-    set spawnerID(id: InstanceID) {
-        if (id) {
+    set spawnerID(id: InstanceID | undefined) {
+        if (id === undefined) {
+            throw new Error("spawnerID cannot be undefined");
+        }
+        if (this.spawnerID === undefined || Buffer.compare(this.spawnerID, id) !== 0) {
```

Comment:
```suggestion
        if (!id.equals(this.spawnerID) {
```

Reply:
OK, I rewrote all of them, supposing for some of them that they can be set to `undefined` or not, depending on the field.